src/internet_identity/src/anchor_management.rs

```diff
@@ -127,23 +146,32 @@ pub async fn update(user_number: UserNumber, device_key: DeviceKey, device_data:
     trap_if_not_authenticated(entries.iter().map(|e| &e.pubkey));
     check_device(&device_data, &entries);
 
-    mutate_device_or_trap(&mut entries, device_key, Some(device_data));
+    let operation = mutate_device_or_trap(&mut entries, device_key, Some(device_data));
 
     write_anchor_data(user_number, entries);
 
     delegation::prune_expired_signatures();
+
+    if state::archive_ready() {
+        archive_operation(user_number, caller(), operation)
+    }
 }
 
 pub async fn remove(user_number: UserNumber, device_key: DeviceKey) {
     let mut entries = state::anchor_devices(user_number);
     // must be called before the first await because it requires caller()
     trap_if_not_authenticated(entries.iter().map(|e| &e.pubkey));
 
+    let caller = caller(); // caller is only available before await
     state::ensure_salt_set().await;
     delegation::prune_expired_signatures();
 
-    mutate_device_or_trap(&mut entries, device_key, None);
+    let operation = mutate_device_or_trap(&mut entries, device_key, None);
     write_anchor_data(user_number, entries);
+
+    if state::archive_ready() {
```

Comment:
this pattern is recurrent, why not call `archive_operation` and let it decide what to do? this will simplify the code, separate concerns, and reduce the risk of calling `archive_operation` if the archive is not ready